File: packages/shared/services/evaluation_storage.py
import logging
import time

from qdrant_client import QdrantClient
from qdrant_client.models import (
    Distance,
    PointStruct,
    VectorParams,
)

from shared.config import config

COLLECTION_NAME = "evaluation_log"
VECTOR_SIZE = config.embedding.vector_size


class EvaluationStorage:
    def __init__(self) -> None:
        self.logger = logging.getLogger(self.__class__.__name__)
        self._client = QdrantClient(host=config.qdrant.host, port=config.qdrant.port)
        self._ensure_collection()

    def _ensure_collection(self) -> None:
        if self._client.collection_exists(collection_name=COLLECTION_NAME):
            return
        self._client.create_collection(
            collection_name=COLLECTION_NAME,
            vectors_config=VectorParams(size=VECTOR_SIZE, distance=Distance.COSINE),
        )
        self.logger.info("Created collection '%s'", COLLECTION_NAME)

    def save(
        self,
        user_input: str,
        agent_output: str,
        tool_calls: list[str],
        score: str,
        feedback: str,
        vector: list[float],
    ) -> None:
        point_id = int(time.time() * 1000)  # ms timestamp as unique id
        self._client.upsert(
            collection_name=COLLECTION_NAME,
            points=[
                PointStruct(
                    id=point_id,
                    vector=vector,
                    payload={
                        "user_input": user_input,
                        "agent_output": agent_output,
                        "tool_calls": tool_calls,
                        "score": score,
                        "feedback": feedback,
                        "created_at": int(time.time()),
                    },
                )
            ],
        )
        self.logger.info("Logged evaluation: score=%s, id=%d", score, point_id)


evaluation_storage = EvaluationStorage()

File: packages/chatbot/server/services/evaluation-storage.ts
import { randomUUID } from "node:crypto";

import { QdrantClient } from "@qdrant/qdrant-js";

import { config, qdrantCollections } from "@/server/config";
import type { EvaluationFeedback } from "@/server/agents/judge";
import { embedderService } from "@/server/services/embedder";

export type EvaluationRecord = {
  id: string;
  userId: string;
  userQuestion: string;
  assistantTrace: string;
  availableTools: string[];
  score: EvaluationFeedback["score"];
  feedback: string;
  criteria: EvaluationFeedback["criteria"];
  issues: EvaluationFeedback["issues"];
  createdAt: number;
};

type SaveEvaluationInput = {
  userId: string;
  userQuestion: string;
  assistantTrace: string;
  availableTools: string[];
  evaluation: EvaluationFeedback;
};

export class EvaluationStorage {
  private static instance: EvaluationStorage | null = null;
  private readonly qdrant: QdrantClient;
  private readonly collectionName = qdrantCollections.evaluation;

  private constructor() {
    this.qdrant = new QdrantClient({ url: config.QDRANT_URL });
  }

  static getInstance(): EvaluationStorage {
    if (!EvaluationStorage.instance) {
      EvaluationStorage.instance = new EvaluationStorage();
    }
    return EvaluationStorage.instance;
  }

  private async ensureCollection(): Promise<void> {
    const { exists } = await this.qdrant.collectionExists(this.collectionName);
    if (!exists) {
      await this.qdrant.createCollection(this.collectionName, {
        vectors: {
          size: config.EMBEDDING_VECTOR_SIZE,
          distance: "Cosine",
        },
      });
    }
  }

  async save({
    userId,
    userQuestion,
    assistantTrace,
    availableTools,
    evaluation,
  }: SaveEvaluationInput) {
    await this.ensureCollection();
    const createdAt = Math.floor(Date.now() / 1000);
    const textToEmbed = [userQuestion, assistantTrace, evaluation.feedback]
      .join("\n")
      .slice(0, 4000);
    const vector = await embedderService.embedQuery(textToEmbed);

    await this.qdrant.upsert(this.collectionName, {
      wait: true,
      points: [
        {
          id: randomUUID(),
          vector,
          payload: {
            userId,
            userQuestion,
            assistantTrace,
            availableTools,
            score: evaluation.score,
            feedback: evaluation.feedback,
            criteria: evaluation.criteria,
            issues: evaluation.issues,
            createdAt,
          },
        },
      ],
    });
  }
}

export const evaluationStorage = EvaluationStorage.getInstance();
